aws-api/src/main/java/com/amplifyframework/api/aws/auth/ApiRequestSignerFactory.java

```diff
@@ -0,0 +1,128 @@
+ */
+
+package com.amplifyframework.api.aws.auth;
+
+import androidx.annotation.NonNull;
+
+import com.amplifyframework.api.ApiException;
+import com.amplifyframework.api.aws.ApiAuthProviders;
+import com.amplifyframework.api.aws.AppSyncGraphQLRequest;
+import com.amplifyframework.api.aws.AuthorizationType;
+import com.amplifyframework.api.aws.sigv4.AppSyncV4Signer;
+import com.amplifyframework.api.graphql.GraphQLRequest;
+import com.amplifyframework.core.Amplify;
+import com.amplifyframework.logging.Logger;
+
+import com.amazonaws.Request;
+
+import java.util.Objects;
+import java.util.function.Supplier;
+
+/**
+ * Factory class that creates instances of different implementations of {@link ApiRequestSigner}s.
+ */
+public final class ApiRequestSignerFactory {
+    private static final Logger LOG = Amplify.Logging.forNamespace("amplify:aws-api");
+    private static final ApiRequestSigner NO_OP_REQUEST_SIGNER = new ApiRequestSigner() {
+        @Override
+        void addAuthHeader(Request<?> request) {
+        }
+    };
+
+    private final ApiAuthProviders apiAuthProviders;
+    private final String region;
+    private final AuthorizationType defaultAuthorizationType;
+
+    /**
+     * Constructor that accepts the API auth providers to be used with their respective request signer.
+     * @param apiAuthProviders An instance with fully configured auth providers for use when signing requests.
+     * @param defaultAuthorizationType The authorization type to use as default.
+     * @param region The AWS region where the API is deployed.
+     */
+    public ApiRequestSignerFactory(@NonNull ApiAuthProviders apiAuthProviders,
+                                   @NonNull AuthorizationType defaultAuthorizationType,
+                                   @NonNull String region) {
+        Objects.requireNonNull(apiAuthProviders);
+        Objects.requireNonNull(defaultAuthorizationType);
+        Objects.requireNonNull(region);
+        this.apiAuthProviders = apiAuthProviders;
+        this.defaultAuthorizationType = defaultAuthorizationType;
+        this.region = region;
+    }
+
+    /**
+     * Return the appropriate request signer after inspecting the request.
+     * @param graphQLRequest The graphQL request sent to the API.
+     * @return The request signer
+     */
+    public ApiRequestSigner fromRequest(GraphQLRequest<?> graphQLRequest) {
+        // If it's not a an instance of AppSyncGraphQLRequest OR
+        // the request's authorization type is null
+        if (!(graphQLRequest instanceof AppSyncGraphQLRequest<?>) ||
+            ((AppSyncGraphQLRequest<?>) graphQLRequest).getAuthorizationType() == null) {
+            // Use the default auth type.
+            return forAuthType(defaultAuthorizationType);
+        } else {
+            // Honor the auth type from the request.
+            return forAuthType(((AppSyncGraphQLRequest<?>) graphQLRequest).getAuthorizationType());
+        }
+    }
```

Comment:
Perhaps slightly cleaner to do something like this?
```suggestion
        AuthorizationType authType = defaultAuthorizationType;
        if(graphQLRequest instanceof AppSyncGraphQLRequest<?>
            && ((AppSyncGraphQLRequest<?>) graphQLRequest).getAuthorizationType() != null) {
            authType = ((AppSyncGraphQLRequest<?>) graphQLRequest).getAuthorizationType()
        }
        return forAuthType(authType);
    }
```